### DISCUSSION TITLE
Getting SGX Measurements of the running enclave during its runtime

### DISCUSSION THREAD

**Discussion by tiagorvmartins** [gramineproject/gramine#1855] [Q&A] [answered]
Hey Gramine team, 

I believe the answer is looking at /dev/attestation/report but is there any kind of library (exposing a function) to use in runtime that could easily fetch and parse the MRENCLAVE and MRSIGNER of the enclave itself running that code?

Been reading about the SGX Report which seems to contain exactly what we need.

On the [low-level interface section documentation](https://gramine.readthedocs.io/en/stable/attestation.html#low-level-dev-attestation-interface):
_/dev/attestation/report pseudo-file can be opened for read and will contain the SGX report._

Thank you in advance!

**Comment by dimakuv** [comment, marked as answer]
Yes, you must read `/dev/attestation/report` binary file.

There is no function, but you can take a look at how Gramine implements the corresponding C functions and do the same/copy-paste them in your C library:
- https://github.com/gramineproject/gramine/blob/a7f46aab10364cfd18565fe803a806405043210a/pal/src/host/linux-sgx/enclave_framework.c#L323
- https://github.com/gramineproject/gramine/blob/a7f46aab10364cfd18565fe803a806405043210a/pal/src/host/linux-sgx/enclave_api.S#L9
- https://github.com/gramineproject/gramine/blob/a7f46aab10364cfd18565fe803a806405043210a/pal/src/host/linux-sgx/enclave_framework.c#L333
- https://github.com/gramineproject/gramine/blob/a7f46aab10364cfd18565fe803a806405043210a/pal/src/host/linux-sgx/enclave_framework.c#L305
- https://github.com/gramineproject/gramine/blob/a7f46aab10364cfd18565fe803a806405043210a/pal/src/host/linux-sgx/sgx_arch.h#L384

Note that before reading the SGX report you should technically write two files:
- `/dev/attestation/user_report_data`
- `/dev/attestation/target_info`

The former may be just all zeros. The latter may be the output of `/dev/attestation/my_target_info`.

Another example: https://github.com/gramineproject/gramine/blob/master/CI-Examples/python/scripts/sgx-report.py

**Reply by tiagorvmartins** [reply]
This is exactly what we need, and thank you for the provided examples!
